src/accessControl/acp/index.js

```diff
@@ -0,0 +1,305 @@
+}
+
+export default class AcpAccessControlStrategy {
+  #datasetWithAcr;
+
+  #policyUrl;
+
+  #fetch;
+
+  constructor(datasetWithAcr, policies, fetch) {
+    this.#datasetWithAcr = datasetWithAcr;
+    this.#policyUrl = getPolicyUrl(datasetWithAcr, policies);
+    this.#fetch = fetch;
+  }
+
+  async getPolicies(policyUrls) {
+    const policyResources = await Promise.all(
+      policyUrls.map((url) => getSolidDataset(url, { fetch: this.#fetch }))
+    );
+    const policies = policyResources
+      .map((resource) => getPolicyAll(resource))
+      .flat();
+    return policies.map((policy) => ({
+      modes: getAllowModesOnPolicy(policy),
+      agents: getRequiredRuleOnPolicy(policy).agents,
+    }));
+  }
+
+  async getPermissions() {
+    const permissions = {};
+    // assigning Read, Write, and Append
+    const accessControls = getAccessControlAll(this.#datasetWithAcr);
+    const policyUrls = accessControls
+      .map((ac) => getPolicyUrlAll(ac).filter((url) => url === this.#policyUrl))
+      .flat();
+    (await this.getPolicies(policyUrls)).forEach(({ modes, agents }) =>
+      agents.forEach((webId) => {
+        const permission = getOrCreatePermission(permissions, webId);
+        permission.acp.apply = addAcpModes(permission.acp.apply, modes);
+        permissions[webId] = permission;
+      })
+    );
+    // assigning Control
+    const controlPolicyUrls = getControlPolicyUrlAll(
+      this.#datasetWithAcr
+    ).filter((url) => url === this.#policyUrl);
+    (await this.getPolicies(controlPolicyUrls)).forEach(({ modes, agents }) =>
+      agents.forEach((webId) => {
+        const permission = getOrCreatePermission(permissions, webId);
+        permission.acp.access = addAcpModes(permission.acp.access, modes);
+        permissions[webId] = permission;
+      })
+    );
+    // normalize permissions
+    return Promise.all(
+      Object.values(permissions).map(async ({ acp, webId }) => {
+        const acl = convertAcpToAcl(acp);
+        return {
+          acl,
+          alias: displayPermissions(acl),
+          profile: await fetchProfile(webId, this.#fetch),
+          webId,
+        };
+      })
+    );
+  }
+
+  async saveApplyPolicyForAgent(dataset, webId, access, mode, acpMap) {
+    const policyUrl = `${this.#policyUrl}#${mode}Policy`;
+    const policy = chain(
+      getOrCreatePolicy(dataset, policyUrl),
+      (p) => setAllowModesOnPolicy(p, acpMap),
+      (p) => setAgents(p, webId, access, mode)
+    );
+    const savedPolicyDataset = await saveSolidDatasetAt(policyUrl, policy);
+    const accessControls = getAccessControlAll(this.#datasetWithAcr);
+    let accessControlWithPolicy = accessControls.find((ac) =>
+      getPolicyUrlAll(ac).find(policyUrl)
+    );
+    if (!accessControlWithPolicy) {
+      accessControlWithPolicy = chain(createAccessControl(), (ac) =>
+        addPolicyUrl(ac, policyUrl)
+      );
+    }
+    this.#datasetWithAcr = setAccessControl(
```

Comment:
Hmm, so this updates `this.#datasetWithAcr` to add the Access Control, but it doesn't actually save it. Which makes me think - I don't think we actually have an API for saving ACRs yet? I've just added a ticket to do so, with the proposed name `saveAccessControlResource`, which would take `this.#datasetWithAcr` as a parameter.

Reply:
Ah wait, I see you're calling a regular `saveSolidDatasetAt()` below, which will make sense if we ever get atomic updates. Until then, though, I think we'll keep them as separate updates like we do for ACLs, to ensure that as a developer, you can deal with either the SolidDataset- or the ACR-update request failing.

Reply:
I've introduced `saveAccessControlResource`, I'll push it with my other updates very soon.